### PR TITLE
Enable file:// protocol in CURL and CCurlFile

### PR DESCRIPTION
**Pull request by basilgello**
## Description

This PR allows distro package builders to add local addon repositories and use them as a fallback if no mirror repository is available or enabled.
Adding the local repo list to official origins allows automatic updates of addons from the local repo by versions from mirrors.kodi.tv repo once it becomes enabled / available.

## Motivation and context

I crafted this change to provide a local repository for language addons on Debian (#25631)

The original problem arose from the following chain of events:

* In 2017, kodi addon repository was made optional to install in Debian [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=877767)
* In 2020, user with intranet-only Kodi setup revealed one can not install any language addon [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1009278)
* In 2023, user complained that 77 language addons I made preinstalled with kodi clutter UX and waste space [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1042996)
* In 2024, I attempted to split language packs into separate Debian packages like firefox or libreoffice do in Debian, but got a REJECT from FTP Masters (#25631)
* To fix this once and for all, I made language packs in a local repo within `/usr/share/kodi/language-repo` and added the local repo to Kodi official repo list to be able to update language packs if mirrors.kodi.tv repo is installed and enabled.

Usage example: https://salsa.debian.org/multimedia-team/kodi-media-center/kodi/-/commit/cc8dcea2ab3d9d375fde8ce0165ccd174bfc46f2  + https://salsa.debian.org/multimedia-team/kodi-media-center/kodi/-/commit/d063166a3e227cd1e2d0e101cbb206cf3c722ba4

## How has this been tested?

1. Install Kodi from Debian 21.1 from unstable
2. Remove kodi-repository-kodi package before first Kodi start
3. Try to set primary UI language different from English
4. See language add-on has been installed and language has been changed

## What is the effect on users?

Users on selected distributions can use repositories bundled inside Kodi distribution package

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: cmake/scripts/common/Macros.cmake [hunk 1]
```diff
@@ -708,6 +708,10 @@ macro(core_find_versions)
     string(TOLOWER ${APP_VERSION_TAG} APP_VERSION_TAG_LC)
   endif()
   string(REPLACE "." "," FILE_VERSION ${APP_ADDON_API}.0)
+  # Let Debian builders etc override APP_ADDON_REPOS
+  if(OVERRIDE_ADDON_REPOS)
+    set(APP_ADDON_REPOS ${OVERRIDE_ADDON_REPOS})
+  endif()
```
Comment (fuzzard) [inline cmake/scripts/common/Macros.cmake:714]: You already patch kodi extensively from what i understand, so why not just patch in your "repo" into https://github.com/xbmc/xbmc/blob/178dc39f30f992291e55a97c582f4cfe2980a248/version.txt#L14

Comment (fuzzard) [reply cmake/scripts/common/Macros.cmake:714]: On this note, does a file:// URI work as is i wonder?
No idea what value your attempting to use to designate the "local" repo.

**Comment by fuzzard** [comment]
In general i highly doubt this is a candidate for a backport

File: cmake/scripts/common/Macros.cmake (hunk 1 above)
Comment (basilgello) [reply cmake/scripts/common/Macros.cmake:714]: Patching version.txt is undesirable as it changes more frequently than CMake code.

file:// works with CCurlFile but not with CFile. To make file:// work, URIUtils::Substitution or sonething must be patched, too

Comment (basilgello) [reply cmake/scripts/common/Macros.cmake:714]: The value for local repo is described in the issue I linked. If Kodi has mirrors.kodi.tv disabled or purely offline / intranet, how should language addons be available? Previous attempt was to install them all but users started complaining about unused language addons keep updating in `$HOME/.kodi/addons` Local repo solves this question quite elegantly.

Comment (fuzzard) [reply cmake/scripts/common/Macros.cmake:714]: I would highly suggest you provide relevant info in the PR description at least. All i see in your link is a half dozen other links.

What is the format you will use to indicate a "local" repo?

Comment (fuzzard) [reply cmake/scripts/common/Macros.cmake:714]: > Patching version.txt is undesirable as it changes more frequently than CMake code.
> 
> file:// works with CCurlFile but not with CFile. To make file:// work, URIUtils::Substitution or sonething must be patched, too

So does file:// mean you dont need to do any of the other changes in this PR? The original code is a CCurlFile after all?

Comment (basilgello) [reply cmake/scripts/common/Macros.cmake:714]: Well, I tested with file:// protocol, had to debug stuff till I found that `file:///usr/share/kodi/language-repo` gets substituted into `file://usr/…` (notice eaten slash between file and usr!), the uri becomes invalid and addon retrieval fails. I switched to special protocol with this change and vooila, it finally worked.

Comment (basilgello) [reply cmake/scripts/common/Macros.cmake:714]: Please note that in order for file:// protocol to start workimg, I also had to patch FetchChecksum into CCurlFile. So:

for special protocol to work, both CRepository::Fetch{Checksum,Index} must be CFile,
for file:// protocol, both CRepository::Fetch{Checksum,Index} must be CCurlFile AND `URIUtils::PerformSubstitution` should be patched notvto transform file protocol links

Comment (basilgello) [reply cmake/scripts/common/Macros.cmake:714]: Updated PR description with relevant info

File: xbmc/addons/Repository.cpp [hunk 2]
```diff
@@ -201,15 +201,25 @@ bool CRepository::FetchIndex(const RepositoryDirInfo& repo,
...
+  // we intentionally avoid using file.GetLength() for
+  // Transfer-Encoding: chunked servers.
+  std::stringstream ss;
+  char temp[1024];
+  int read;
+  while ((read = file.Read(temp, sizeof(temp))) > 0)
+    ss.write(temp, read);
+  if (read <= -1)
+    return false;
+  response = ss.str();
```
Comment (neo1973) [inline xbmc/addons/Repository.cpp:221]: To avoid copying the string into `response`:
```suggestion
  response = std::move(ss).str();
```

**Comment by basilgello** [comment]
@fuzzard Without the backport label, do you still think this PR brings insufficient value to the project?

**Comment by fuzzard** [comment]
My honest opinion is im not a fan of the override. I think overriding the version.txt repo list is just risky, and my opinion, as much as im sure you're not going to like this, is that the system repos should be taken in via the version.txt as a global source of truth for a build.
Regarding the actual use of a special://, i think thats a hack as well, and i think a better solution would be to support the file:// curl URI properly from the version.txt file.

Im not going to stand in the way if someone else wants to give it the go ahead, but i think theres a better solution possible.

**Comment by basilgello** [comment]
Thanks for the honest opinion! I think file:// protocol is doable as well as patching version.txt (not by means of a patchfile but a sed step in debian/rules because patchfile is too brittle for a file that is updated each point release!). I just could not solve path substitution bug in a very compressed timeline. I hope someone more familiar with that internals can point how to implement it better.

**Comment by basilgello** [comment]
If I understand the "expected" solution properly, version.txt should have `file:///usr/share/kodi/language-repo` instead of `special://xbmc` ?

**Comment by fuzzard** [comment]
Yeah, that's how I would think it should be.

**Comment by basilgello** [comment]
OK, let me debug that path substitution issue transforming `file:///usr/share/kodi/language-repo` into `file://usr/share/kodi/language-repo` - maybe the easiest solution is just to exclude `file://` protocol from `CURIUtils::SubstitutePath`

**Comment by basilgello** [comment]
@fuzzard Refactored as per your suggestion!

File: xbmc/addons/Repository.cpp (hunk 2 above)
Comment (basilgello) [reply xbmc/addons/Repository.cpp]: @neo1973 I solved the problem via different set of files so your `std::move` suggestions have fallen out of scope for this PR.

But I have prepared a new PR just for that: #25728

**Comment by basilgello** [comment]
Not sure if it is worth adding test for `file://` protocol to TestURIUtils

**Comment by basilgello** [comment]
@fuzzard Marked this as ready for review :)

**Review by garbear — APPROVED** [review]
I missed the original version of the PR and its discussion but the current state looks good. The documentation for the added "file" check states:

```
// they are all local protocols and have no server part, port number, special options, etc.
```

As `file://` has no "server", the change here looks correct.

**Comment by basilgello** [comment]
Thanks for approval @garbear ! You missed nothing as such - I made a solution for my targeted usecase in Debian and fuzzard pointed out that more generic solution would be better for the project :)

If no one has objections, I will merge this on Monday.

File: xbmc/filesystem/CurlFile.cpp [hunk 3]
```diff
@@ -1097,7 +1097,7 @@ bool CCurlFile::Open(const CURL& url)
 
   m_httpresponse = m_state->Connect(m_bufferSize);
 
-  if (m_httpresponse <= 0 || (m_failOnError && m_httpresponse >= 400))
+  if (m_httpresponse < 0 || (m_failOnError && m_httpresponse >= 400))
```
Comment (fuzzard) [inline xbmc/filesystem/CurlFile.cpp:1100]: My concern with this is the history. https://github.com/xbmc/xbmc/pull/7290
it was specifically added to test specifically 0 error code due to things potentially not returning.

I dont know why this is required for this PR (i havent looked into it), but i think another solution may be necessary to take the original intent/use into consideration

Comment (basilgello) [reply xbmc/filesystem/CurlFile.cpp:1100]: curl returns HTTP response code 0 for file protocol. Maybe worth adding check allowing 0 for file:// specifically ?

Comment (fuzzard) [reply xbmc/filesystem/CurlFile.cpp:1100]: yeah, thats one way i had thought might be viable. havent looked any further into surrounding code though

Comment (basilgello) [reply xbmc/filesystem/CurlFile.cpp]: OK, added the guardrail for file protocol!

**Comment by basilgello** [comment]
jenkins build this please

**Pull request merged**